### DISCUSSION TITLE
sanitize_name and field name starting by number

### DISCUSSION THREAD

**Discussion by JaSei** [bxparks/bigquery-schema-generator#84] [General]
Hi folks.
First - your tool works great thanks for it.
Unfortunately, the data I work with them is a mess. What I am fighting with now is this JSON

```
{"objects":{"0":{"mime_type":"application/octet-stream","type":"artifact","hashes":{"MD5":"6..1","SHA-1":"4..0","SHA-256":"4..f"},"url":"https://URL/artifacts/4..f","x_cta_hash_identity":"6..a","x_cta_hash_context":"2..2","spec_version":"2.0"}}}
```
as you see, there is the map, which key is named as number `0`... but BigQuery doesn't support numbers as the first letter

```
BigQuery error in load operation: Invalid field name "0". Fields must contain only letters, numbers, and underscores, start with a letter or underscore, and be at most 300 characters long.
```

So I propose to change `"0"` to `"_0"` in this case when `--santizie_name` is applied.

Thanks

**Comment by JaSei** [comment]
something like this?
`new_value = re.sub('^(\d)', '_\1', new_value)` in the

```
    def sanitize_name(self, value):
        """Sanitizes a column name within the schema.

        We explicitly choose to not perform the lowercasing here as this
        cause us to lose case sensitivity when generating the final schema
        """
        if self.sanitize_names:
            new_value = re.sub('[^a-zA-Z0-9_]', '_', value[:127])
            new_value = re.sub('^(\d)', '_\1', new_value)
        else:
            new_value = value
        return new_value
```

**Reply by bxparks** [reply]
I would prefer avoid doing data cleansing inside bigquery-schema-generator. Each user will want to convert the illegal column names into legal names in different ways. The correct transformation depends on the context of the input data, and I don't want to be the repository of all the different ways to clean up the corrupted data for all the different downstream projects that use this tool. So my recommendation is for you to write your own data cleansing script that's appropriate for your project.

**Reply by bxparks** [reply]
You may be wondering why `--sanitize_name` is supported. It's because the last time I checked, the `bq load --autodetect` command performs that particular sanitization, but for CSV files only. So bigquery-schema-generator tries to emulate that. For ND-JSON files, `bq load` considers illegal column names as an error and spit out an error message. Technically, I guess bigquery-schema-generator should throw an exception if `--sanitize_name` is given for an ND-JSON file. But I didn't write the code that implements it, and I thought it would be too pedantic to impose that restriction, so I just let it go into the code base.
